### PR TITLE
[flang] Make `bbc`, `tco` and `flang-new` share pass pipeline creation

### PR DESCRIPTION
**Pull request by banach-space**
As noted in [1], after [2] `flang-new -fc1 -emit-llvm` started failing
for a sample input source file. That's because a pass pipeline for
simplifying FIR was modified inconsistently in `bbc` and in
CLOptions.inc. This patch fixes that inconsistency and makes sure that
`bbc`, `tco` and `flang-new` share the code for generating pass
pipelines.

In summary:
* `createMLIRToLLVMPassPipeline` is reduced and renamed as
  `createBasicFirPassPipeline` (so that it can be shared with `bbc`)
* passes for generating code from `createMLIRToLLVMPassPipeline` are
  extracted into `addCodeGenPasses` (only to be used in `tco` and
  `flang-new`)
* the pass pipeline in `bbc` is not changed, the pass pipelines in `tco`
  and `flang-new` are made consistent with `bbc`

[1] https://github.com/flang-compiler/f18-llvm-project/issues/1280
[2] https://github.com/flang-compiler/f18-llvm-project/pull/1126

### CODE REVIEW

File: flang/include/flang/Tools/CLOptions.inc [hunk 1]
```diff
@@ -70,40 +70,16 @@ inline void addCSE(mlir::PassManager &pm) {
...
-  addPassConditionally(pm, disableLlvmIrToLlvm,
-      [&]() { return fir::createLLVMDialectToLLVMPass(output); });
-}
-
-/// Create a pass pipeline for lowering from MLIR to LLVM IR
+/// Create a basic pass pipeline for simplyfing the FIR dialect
 ///
 /// \param pm - MLIR pass manager that will hold the pipeline definition
-inline void createMLIRToLLVMPassPipeline(mlir::PassManager &pm)
+inline void createBasicFirPassPipeline(mlir::PassManager &pm)
```
Comment (schweitzpgi) [inline flang/include/flang/Tools/CLOptions.inc:76]: This name change obfuscates what this pass pipeline actually does and effectively reverts a name change request made during upstreaming efforts. The pipeline takes MLIR and converts it to LLVM.

File: flang/include/flang/Tools/CLOptions.inc [hunk 2]
```diff
@@ -70,40 +70,16 @@ inline void addCSE(mlir::PassManager &pm) {
       pm, disableFirCse, fir::createCSEPass);
 }
 
-#if !defined(FLANG_EXCLUDE_CODEGEN)
```
Comment (schweitzpgi) [inline flang/include/flang/Tools/CLOptions.inc:73]: Why are these moved? Seems like gratuitous change for the sake of change.

File: flang/include/flang/Tools/CLOptions.inc [hunk 3]
```diff
@@ -70,40 +70,16 @@ inline void addCSE(mlir::PassManager &pm) {
...
+inline void createBasicFirPassPipeline(mlir::PassManager &pm)
 {
   // simplify the IR
-  mlir::GreedyRewriteConfig config;
-  config.enableRegionSimplification = false;
   fir::addCSE(pm);
   pm.addNestedPass<mlir::FuncOp>(fir::createArrayValueCopyPass());
   pm.addNestedPass<mlir::FuncOp>(fir::createCharacterConversionPass());
-  pm.addPass(mlir::createCanonicalizerPass(config));
+  pm.addPass(mlir::createCanonicalizerPass());
```
Comment (schweitzpgi) [inline flang/include/flang/Tools/CLOptions.inc:82]: This looks incorrect. Running MLIR's full canonicalizer pass deploys features that break the compiler.

File: flang/include/flang/Tools/CLOptions.inc [hunk 4]
```diff
@@ -117,10 +93,38 @@ inline void createMLIRToLLVMPassPipeline(mlir::PassManager &pm)
   pm.addNestedPass<mlir::FuncOp>(fir::createControlFlowLoweringPass());
   pm.addPass(mlir::createLowerToCFGPass());
 
+  mlir::GreedyRewriteConfig config;
+  config.enableRegionSimplification = false;
   pm.addPass(mlir::createCanonicalizerPass(config));
```
Comment (schweitzpgi) [inline flang/include/flang/Tools/CLOptions.inc:98]: Why use a config here but not above?

File: flang/tools/bbc/bbc.cpp [hunk 5]
```diff
@@ -267,30 +267,7 @@ static mlir::LogicalResult convertFortranSourceToMLIR(
...
-    fir::addCfgConversionPass(pm);
-    pm.addNestedPass<mlir::FuncOp>(fir::createControlFlowLoweringPass());
-    pm.addPass(mlir::createLowerToCFGPass());
-
-    mlir::GreedyRewriteConfig config;
-    config.enableRegionSimplification = false;
-    pm.addPass(mlir::createCanonicalizerPass(config));
-    pm.addPass(fir::createSimplifyRegionLitePass());
-    fir::addCSE(pm);
+    fir::createBasicFirPassPipeline(pm);
```
Comment (schweitzpgi) [inline flang/tools/bbc/bbc.cpp:270]: This seems wrong. `bbc` doesn't convert MLIR to LLVM.

**Review by schweitzpgi — CHANGES_REQUESTED** [review]

File: flang/include/flang/Tools/CLOptions.inc (hunk 1 above)
Comment (banach-space) [reply flang/include/flang/Tools/CLOptions.inc:76]: Only `createBasicFirPassPipeline` is moved. That's required in order for `bbc` to be able to use it. Methods for code-gen are [excluded](https://github.com/flang-compiler/f18-llvm-project/blob/fir-dev/flang/tools/bbc/bbc.cpp#L133-L134) in `bbc`. However, with this change `createBasicFirPassPipeline` becomes a method that we do want to use in `bbc` (I removed code-gen passes from there).

The diff suggests that other methods are moved as well, but that's how `git diff` is displayed. There isn't much I can do about this.

File: flang/include/flang/Tools/CLOptions.inc (hunk 2 above)
Comment (banach-space) [reply flang/include/flang/Tools/CLOptions.inc:73]: Only `createBasicFirPassPipeline` is moved. That's required in order for `bbc` to be able to use it. Methods for code-gen are [excluded](https://github.com/flang-compiler/f18-llvm-project/blob/fir-dev/flang/tools/bbc/bbc.cpp#L133-L134) in `bbc`. However, with this change `createBasicFirPassPipeline` becomes a method that we do want to use in `bbc` (I removed code-gen passes from there).

The diff suggests that other methods are moved as well, but that's how `git diff` is displayed. There isn't much I can do about this.

File: flang/include/flang/Tools/CLOptions.inc (hunk 1 above)
Comment (banach-space) [reply flang/include/flang/Tools/CLOptions.inc:76]: > This name change obfuscates what this pass pipeline actually does

Please, could you suggest a more suitable change?

> effectively reverts a name change request made during upstreaming efforts

I don't believe that's the case. I introduced this method in https://github.com/flang-compiler/f18-llvm-project/pull/1113 to share code with `tco` and `flang-new`. Now it seems like a good time to refactor it so that it can be shared with `bbc` too. Also, AFAIK, this method is not present upstream _yet_?

File: flang/include/flang/Tools/CLOptions.inc (hunk 3 above)
Comment (banach-space) [reply flang/include/flang/Tools/CLOptions.inc:82]: This was copied from `bbc` (so that the code can be shared between `bbc` and `tco`/`flang-new`). Is `bbc` incorrect then? (extracted from [here](https://github.com/flang-compiler/f18-llvm-project/blob/fir-dev/flang/tools/bbc/bbc.cpp#L270-L294)). Or are `bbc` and `tco`/`flang-new` meant to diverge here?

File: flang/include/flang/Tools/CLOptions.inc (hunk 4 above)
Comment (banach-space) [reply flang/include/flang/Tools/CLOptions.inc:98]: Tl;Dr I don't know, I'm not the original author.

This code was extracted from `bbc` (from [here](https://github.com/flang-compiler/f18-llvm-project/blob/fir-dev/flang/tools/bbc/bbc.cpp#L270-L294)). This code in `bbc` was recently refactored in https://github.com/flang-compiler/f18-llvm-project/pull/1126.

@clementval ?

File: flang/tools/bbc/bbc.cpp (hunk 5 above)
Comment (banach-space) [reply flang/tools/bbc/bbc.cpp:270]: I'm guessing that you find this method name not suitable? What would be a good alternative?

**Review by banach-space — COMMENTED** [review]
@schweitzpgi Thanks for taking a look! I've replied inline.

FYI, you may also want to take a look at https://github.com/flang-compiler/f18-llvm-project/issues/1280 that this PR is trying to fix.

**Comment by schweitzpgi** [comment]
See #1287 for a corrected version of #1126.

**Comment by banach-space** [comment]
Closing in favor of #1287 and [ Refactor creating the pass pipeline so the optimizer portion can be shared with bbc.](https://github.com/flang-compiler/f18-llvm-project/commit/a2169ecfae5c1fde9e81bc4d346d6362c0d597d8)
